### ISSUE TITLE
Orca does not read a non letter character on a line

### ISSUE THREAD

**Issue by jvesouza** [microsoft/vscode#93091]
Issue Type: <b>Bug</b>

1. Create a json file with the following content:
>{
>    "alpha": {
>        "a": "a"
>    }
>}
2. Activate orca.
3. Using vscode open the file created in the step 1.
4. Move the cursor to the second line.
5. Make sure that the cursor is in the first position of the line, column 1.
6. Press shift+down twice.
7. Press shift+down one more time.

Expected result:
Orca should announce the contents of line 4 as selected. This line contains a }.

Result received:
Orca announces blank, as if the line has no content.

VS Code version: Code - Insiders 1.44.0-insider (64b65cf8e401cb98d4ddfc8324cb6635ff6fc2ed, 2020-03-20T08:27:32.236Z)
OS version: Linux x64 5.5.10-arch1-1

@isidorn @ajborka Would be possible to confirm? Thanks.

**Comment by isidorn** [comment]
@jvesouza yes I can reproduce this issue.
The problem is not selection, but that Orca does not read the first non letter character on a line. Thus renaming issue.
Have the following content:
```
a
}
;
```
Have the cursor on start of each line, move down and notice how only `a` is being read out, `}` and `;` will not get read out by orca.
This is either a Chrome issue or an Orca issue.
@joanmarie thoughts? I can create a standalone Chrome repro so we file an issue against chrome?

**Comment by joanmarie** [comment]
This might be an Orca issue. Lemme look.

**Comment by joanmarie** [comment]
> The problem is not selection, but that Orca does not read the first non letter character on a line. Thus renaming this issue.
> Have the following content:
> 
> ```
> a
> }
> ;
> ```
> 
> Have the cursor on start of each line, move down and notice how only `a` is being read out, `}` and `;` will not get read out by orca.

Do you have punctuation set to all in Orca's preferences? If not, then you will see the behavior you describe and it doesn't matter what your editor is (e.g. Gedit).

It's been on my todo list to ignore the user's punctuation preferences when the only thing on the line is punctuation to force the punctuation to be spoken rather than silence.

While still debugging the original report from @jvesouza, I believe that it is an Orca issue and related to selection.

**Comment by isidorn** [comment]
@joanmarie I have all the default Orca preferences. 
Yes not respecting that setting when punctuation is the only thing on a line makes sense.
Also is some way for vscode to tell orca (other screen readers) that punctuation should be always read. Since VS Code is a code editor and every character insde the editor matters.

Then I guess @jvesouza steps are the correct one. 
@joanmarie Let me know if it is not an Orca issue and we can look into creating a chrome issue.

For chrome issues in general note that VS Code uses a hidden HTML `textarea`, and thus whenever i try to repro issues in Chrome I just take a page with a simple textarea. For example this one https://bocupp-microsoft.github.io/sample-code/textarea/setselectionrange-performance.html

**Comment by joanmarie** [comment]
> Yes not respecting that setting when punctuation is the only thing on a line makes sense.

That's my gut reaction, but I have been known to be wrong before. :grinning: So I just asked on the Orca list. I'll hold off until users chime in.

> Also is some way for vscode to tell orca (other screen readers) that punctuation should be always read. Since VS Code is a code editor and every character insde the editor matters.

Funny you mention that. We recently added a new `code` role to ARIA whose purpose is to accomplish that. https://w3c.github.io/aria/#code. I've not yet added support for it in Orca, but this conversation suggests perhaps I should prioritize doing so.

Also at the time that role was added, I don't think we were giving much thought to whether or not something with that role might be editable or not. I don't see why it couldn't be though. That said, before using it, we'd of course want to be sure that it doesn't break anything for any screen reader. In other words, even for screen readers that didn't yet support the `code` role, do they continue to work with that role applied to the editor?

**Comment by isidorn** [comment]
@joanmarie great, let's wait and see the user feedback.

As for the role `code` I think it makes sense to add it and thanks for providing a pointer. Since screen readers could do quite some optimisations for the Code Editor use case.
And yes we can not jump on it until all screen readers support this. If Orca starts supporting this before others we can of course put that role only when we are running on Linux.
The current role we use is `textbox`. And if we changed to `code` for all screen readers that do not yet support this, it would probably lead to issues.

**Comment by joanmarie** [comment]
The original bug reported by @jvesouza has been fixed in Orca master as well as the stable (gnome-3-36) branch.

**Comment by isidorn** [comment]
Thanks for the quick fix!
Thus closing this.
As for the `code` role just ping me on a new issue if orca adopts it and we can do the change in VS Code.
